opencog/distvalue/DVecUtils.cc

```diff
@@ -0,0 +1,195 @@
+#include <algorithm>
+#include <cmath>
+#include <sstream>
+#include <ostream>
+
+#include <opencog/util/exceptions.h>
+#include <opencog/util/numeric.h>
+
+#include <opencog/distvalue/DVecUtils.h>
+
+namespace opencog
+{
+
+DVec operator+(const DVec& a, const DVec& b)
+{
+    DVec result;
+
+    const std::size_t n = std::min(a.size(), b.size()) ;
+    std::transform(std::begin(a), std::begin(a)+n, std::begin(b),
+                   std::back_inserter(result), std::plus<double>{});
+    return result;
+}
+
+DVec operator-(const DVec& a, const DVec& b)
+{
+    DVec result;
+
+    const std::size_t n = std::min(a.size(), b.size()) ;
+    std::transform(std::begin(a), std::begin(a)+n, std::begin(b),
+                   std::back_inserter(result), std::minus<double>{});
+    return result;
+}
+
+DVec operator*(const DVec& a, const DVec& b)
+{
+    DVec result;
+
+    const std::size_t n = std::min(a.size(), b.size()) ;
+    std::transform(std::begin(a), std::begin(a)+n, std::begin(b),
+                   std::back_inserter(result), std::multiplies<double>{});
+    return result;
+}
+
+DVec operator/(const DVec& a, const DVec& b)
+{
+    DVec result;
+
+    const std::size_t n = std::min(a.size(), b.size()) ;
+    std::transform(std::begin(a), std::begin(a)+n, std::begin(b),
+                   std::back_inserter(result), std::divides<double>{});
+    return result;
+}
+
+DVec operator*(const DVec & a, double b)
+{
+    DVec result;
+
+	for (double elem : a)
+		result.push_back(elem * b);
+
+    return result;
+}
+
+DVec operator/(const DVec & a, double b)
+{
+    DVec result;
+
+	for (double elem : a)
+		result.push_back(elem / b);
+
+    return result;
+}
+
+bool operator<(const DVec & a, const DVec & b)
+{
+	return sum(a) < sum(b);
+}
+
+bool operator>(const DVec & a, const DVec & b)
+{
+	return sum(a) > sum(b);
+}
+
+bool operator==(const DVec & a, const DVec & b)
+{
+	if (a.size() != b.size())
+		return false;
+	for (unsigned int i = 0; i < a.size(); i++)
+	{
+		if (!is_approx_eq_ulp(a[i],b[i],24))
+			return false;
+	}
+	return true;
+}
+
+double sum(const DVec & a)
+{
+	double res = 0;
+	for (double e : a)
+		res += e;
+	return res;
+}
+
+double dist(const DVec & p1, const DVec & p2)
+{
+	double dist = 0;
+    for(size_t i = 0; i < p1.size(); i++)
+    {
+        double d = p1[i] - p2[i];
+        dist += d * d;
+    }
+    return sqrt(dist);
+}
+
+double dot(const DVec & a, const DVec & b)
+{
+	if (a.size() != b.size())
+		throw RuntimeException(TRACE_INFO,"Vectors must be the same lenght.");
+	double sum = 0;
+	for (size_t i = 0; i < a.size(); i++)
+	{
+		sum += a[i] * b[i];
+	}
+	return sum;
```

Comment:
You could use https://en.cppreference.com/w/cpp/algorithm/inner_product or https://www.boost.org/doc/libs/1_71_0/libs/range/doc/html/range/reference/algorithms/numeric/inner_product.html